yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryProvider.scala

```diff
@@ -0,0 +1,1015 @@
+   * if it is enabled then the URL is logged too.
+   */
+  private def init(): Unit = {
+    if (!enabled) {
+      logError(YarnHistoryProvider.TEXT_SERVICE_DISABLED)
+    } else {
+      logInfo(YarnHistoryProvider.TEXT_SERVICE_ENABLED)
+      logInfo(YarnHistoryProvider.KEY_SERVICE_URL + ": " + timelineEndpoint)
+      logDebug(sparkConf.toDebugString)
+      // get the thread time
+      logInfo(s"refresh interval $refreshInterval milliseconds")
+      if (refreshInterval < 0) {
+        throw new Exception(YarnHistoryProvider.TEXT_INVALID_UPDATE_INTERVAL +
+            s": ${refreshInterval/1000}")
+      }
+      startRefreshThread()
+    }
+  }
+
+
+  /**
+   * Stop the service. After this point operations will fail.
+   */
+  override def stop(): Unit = {
+    logDebug(s"Stopping $this")
+    // attempt to stop the refresh thread
+    if (!stopRefreshThread()) {
+      closeQueryClient()
+    }
+
+  }
+
+  /**
+   * Close the query client
+   */
+  def closeQueryClient(): Unit = {
+    // and otherwise, stop the query client
+    logDebug("Stopping Timeline client")
+    timelineQueryClient.close()
+  }
+
+  /**
+   * Is the timeline service (and therefore this provider) enabled.
+   * (override point for tests).
+   *
+   * Important: this is called during construction, so test-time subclasses
+   * will be invoked before their own construction has taken place.
+   * Code appropriately.
+   * @return true if the provider/YARN configuration enables the timeline
+   *         service.
+   */
+  def enabled: Boolean = {
+    _enabled
+  }
+  
+  /**
+   * Get the timeline query client. Used internally to ease testing
+   * @return the client.
+   */
+  def getTimelineQueryClient(): TimelineQueryClient = {
+    timelineQueryClient
+  }
+
+  /**
+   * Set the last exception
+   * @param ex exception seen
+   */
+  private def setLastFailure(ex: Throwable): Unit = {
+    setLastFailure(ex, now())
+  }
+
+  /**
+   * Set the last exception
+   * @param ex exception seen
+   * @param timestamp the timestamp of the failure
+   */
+  private def setLastFailure(ex: Throwable, timestamp: Long): Unit = {
+    this.synchronized {
+      lastFailureCause = Some(ex, new Date(timestamp))
+    }
+  }
+
+  /**
+   * Reset the failure info
+   */
+  private def resetLastFailure(): Unit = {
+    this.synchronized {
+      lastFailureCause = None
+    }
+  }
+
+  /**
+   * Get the last exception
+   * @return the last exception or  null
+   */
+  def getLastFailure(): Option[(Throwable, Date)] = {
+    this.synchronized {
+      lastFailureCause
+    }
+  }
+
+  /**
+   * Query for the connection being healthy
+   * @return
+   */
+  def isHealthy(): Boolean = {
+    healthy.get()
+  }
+
+  /**
+   * Get that the health flag itself. This allows test code to initialize it properly.
+   * Also: if accessed and set to false, it will trigger another health chek.
+   * @return
+   */
+  protected def getHealthFlag(): AtomicBoolean = {
+    healthy;
```

Comment:
Still, my main comment was about the name of the variable being misleading.

If your goal is to have just some health check, why not do it off-band? Why do it inline with the requests and add all this complexity? Just have a thread that periodically tries to connect to the ATS. It would have the same behavior as this code, since right now the flag does not prevent user requests from going forward when the ATS is unhealthy anyway.